gcc/rust/rust-session-manager.h

```diff
@@ -205,9 +208,31 @@ struct CompileOptions
     enable_dump_option (DumpOption::TYPE_RESOLUTION_DUMP);
   }
 
+  /* Validate the crate name using the ASCII rules
+   TODO: Support Unicode version of the rules */
+  bool validate_crate_name (const std::string crate_name)
+  {
+    if (crate_name.length () > kMaxNameLength || crate_name.empty ())
+      return false;
+    for (auto &c : crate_name)
+      {
+	if (!(ISALNUM (c) || c == '_' || c == '-'))
+	  return false;
+      }
+    return true;
+  }
+
   bool set_crate_name (std::string name)
   {
-    // TODO: validate the crate name?
+    if (!validate_crate_name (name))
+      return false;
+
+    /* Replace all the '-' symbols with '_' per Rust rules */
+    for (size_t i = 0; i < name.size (); i++)
+      {
+	if (name[i] == '-')
+	  name[i] = '_';
+      }
```

Comment:
```suggestion
    for (auto &c: name)
	if (c == '-')
	  c = '_';
```
Any reason why you use the range based for loop above and not here?

Reply:
> Any reason why you use the range based for loop above and not here?

I had an incorrect intuition that C++ for-range loops can't mutate the elements, so I used the C-style for loop here.

I will switch to the for-range loop here

Reply:
I'm pretty sure the rules are as follow (I'm not well-versed in C++ by any means):

1. `auto &elt` is just a regular reference, so it is mutable by default
2. `const auto &elt` is a const reference, immutable
3. `auto elt` creates a copy of the element, so the copy is mutable but will not modify the underlying iterator
4. `const auto elt` also create a copy, but an immutable one obviously